components/omega/src/ocn/OceanInit.cpp

```diff
@@ -0,0 +1,217 @@
+//===-- ocn/OceanInit.cpp - Ocean Initialization ----------------*- C++ -*-===//
+//
+// This file contians ocnInit and associated methods which initialize Omega.
+// The ocnInit process reads the config file and uses the config options to
+// initialize time management and call all the individual initialization
+// routines for each module in Omega.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AuxiliaryState.h"
+#include "Config.h"
+#include "DataTypes.h"
+#include "Decomp.h"
+#include "Field.h"
+#include "Halo.h"
+#include "HorzMesh.h"
+#include "IO.h"
+#include "Logging.h"
+#include "MachEnv.h"
+#include "OceanDriver.h"
+#include "OceanState.h"
+#include "TendencyTerms.h"
+#include "TimeMgr.h"
+#include "TimeStepper.h"
+
+#include "mpi.h"
+
+namespace OMEGA {
+
+int ocnInit(MPI_Comm Comm,          ///< [in] ocean MPI communicator
+            Calendar &OmegaCal,     ///< [out] sim calendar
+            TimeInstant &StartTime, ///< [out] sim start time
+            Alarm &EndAlarm         ///< [out] alarm to end simulation
+) {
+
+   // Error codes
+   I4 RetErr = 0;
+   I4 Err    = 0;
+
+   // Init the default machine environment based on input MPI communicator
+   MachEnv::init(Comm);
+   MachEnv *DefEnv = MachEnv::getDefault();
+
+   // Initialize Omega logging
+   initLogging(DefEnv);
+
+   // Read config file into Config object
+   Config("omega");
+   Err = Config::readAll("omega.yml");
+   if (Err != 0) {
+      LOG_ERROR("ocnInit: Error reading config file");
+      ++RetErr;
+   }
+   Config *OmegaConfig = Config::getOmegaConfig();
```

Comment:
I'm wanting to understand better the liberal use of `LOG_ERROR()` vs. `LOG_CRITICAL()`.  Maybe this is a philosophical question for Omega more broadly.

I personally see relatively little value in accumulating a bunch of errors but only terminating execution when they have knock-on effects that terminate the code.  If an error has occurred and it is not fatal, that is not an error, that is a warning.

The MPAS framework used non-fatal error messages but this has not been something the E3SM team is happy with in general.  An error message means something is wrong and needs to be fixed.  It feels to me like there is little benefit in continuing the code in such cases.

@brian-oneill, @philipwjones, @sbrus89 and the Omega team more broadly, what do you all see as the value of non-critical errors (outside of the testing infrastructure, perhaps)?

Reply:
@xylar You are correct in saying we've been pretty loose in categorizing errors. There are a couple of aspects to this.  First is how we label a message while the second is the behavior we want in each case. The canonical use of error vs critical is that error signifies an error that causes a function to fail or give a wrong answer but that the overall application could proceed, while critical means that there is an error in a critical function for the application that will cause it to fail. We should certainly be a little more rigorous in how we label the messages. (Note that the difference between error and warn is that error is really the wrong answer or a true error while warn refers to a change in behavior that the user might not have expected/desired - like using a default value when a config is missing)

In terms of behavior, we could technically keep running with an error (based on the definition above), but I think in nearly all cases, we want Omega to exit anyway rather than continue running with a bad solution. So there is little practical difference in the behavior we want between Error and Critical most of the time. Our current spdlog behavior is the same in both cases - both emit to std error/log file and flush I think? We can certainly change that if needed though see my next comment. Even with identical behavior, we should still label messages appropriately to more correctly assess/describe their severity.

One other aspect of behavior that I want to retain - I'd rather most routines pass the error/critical back up the call tree so we can get a trace. Ideally, control would return all the way to the top-level driver where a cleaner exit/abort could take place, though this is not always possible.  And we don't always want error checks in the middle of performance-critical code where the interrupts/check could cause slowdowns.

Reply:
@philipwjones, I could be wrong but I would have thought that c++ had ways of doing stack traces without the need for passing the error all the way back up to the driver.  That has been one of the most poorly behaved parts of the MPAS framework and components in my view, and Omega seems to be replicating the bad behavior.  If there is a way for c++ to produce a stack trace in the `LOG_CRITICAL()` function rather than relying on passing the error code back up the stack to the driver, I would think that would be much less fragile and prone to lazy developer practices.

It sounds like we should have a discussion about this at today's or a future Omega meeting.